### ISSUE TITLE
Unevenly spaced ticks in plot_date()

### ISSUE THREAD

**Issue by chrisjbillington** [matplotlib/matplotlib#13183]
### Bug report

When plotting with `plot_date()`, the x axis ticks can be unevenly spaced. It looks like when the tick spacing is less than a month, the start of the month is always present as a tick, which may mean the space between it and the previous tick is small. This can lead to overlapping tick labels and generally poor looking axes and grids.

**Code for reproduction**

```python
import numpy as np
import matplotlib.pyplot as plt

start = np.datetime64('2018-01-01', 'D')
end = np.datetime64('2018-04-01', 'D')
dates = np.arange(start, end)
data = np.random.randn(len(dates))

plt.plot_date(dates, data)
plt.grid(True)
plt.gcf().autofmt_xdate()
plt.show()
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/1044087/51126029-11384100-17f0-11e9-8341-28790dc7d8e5.png)

**Expected outcome**

The plot should have evenly spaced gridlines and axis ticks, unless a custom locator has specified otherwise.

**Matplotlib version**

  * Operating system: Arch Linux
  * Matplotlib version: 3.0.2
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.7.2
  * Jupyter version (if applicable): n/a
  * Other libraries: matplotlib 1.15.4

Matplotlib, Python and numpy are installed with the OS package manager using the standard Arch Linux repositories.

**Comment by jklymak** [comment]
Yes, this was an unintended consequence of #9801.  Its been "fixed" by #10841 which is on master...

![figure_1](https://user-images.githubusercontent.com/1562854/51130452-e1823c80-17e1-11e9-8436-fe322864fccc.png)

Now "fixed" is still different to how it was before, which was (v2.2.3):

![figure_1-3](https://user-images.githubusercontent.com/1562854/51130515-037bbf00-17e2-11e9-9082-93aae92a681c.png)

If you want the old behaviour back, you need to do a bit of a song and dance.  

```python
locator = mdates.AutoDateLocator(interval_multiples=False)
formatter = mdates.AutoDateFormatter(locator)

plt.plot_date(dates, data)
plt.gca().xaxis.set_major_locator(locator)
plt.gca().xaxis.set_major_formatter(formatter)

plt.grid(True)
plt.gcf().autofmt_xdate()
plt.show()
```

I agree that the way things are in 3.0.2 are undersirable because to intervals were not good (i.e. you could get 29, 1 ticks).  OTOH, I think the new behaviour (as of master) is better than the 2.2.3 behaviour.

**Comment by chrisjbillington** [comment]
Oh, that sounds like a great improvement! Snapping the ticks to month edges makes a lot of sense.

I have no complaints then, and will close this issue since it's fixed already.

**Comment by jklymak** [comment]
No problem - sorry for the inconvenience of the bad state right now...

**Comment by jklymak** [comment]
Oh, sorry, "evenly spaced" for months won't work, of course....  Note the ticks are *not* evenly spaced above...

**Comment by chrisjbillington** [comment]
Oh, right, since not all months have the same number of days. Well, it's close enough not to be jarring and is more understandable than exactly even spacing (I guess I didn't realise prior to now that this was a trade-off, I just though matplotlib had been choosing the ticks poorly for no reason).
